src/VisualStudio/Core/Def/DocumentOutline/DocumentOutlineControl.xaml.cs

```diff
@@ -0,0 +1,291 @@
+
+        /// <summary>
+        /// Keeps track of the current primary and secondary text views. Should only be accessed by the UI thread.
+        /// </summary>
+        private readonly Dictionary<IVsTextView, ITextView> _trackedTextViews = new();
+
+        public DocumentOutlineControl(
+            ILanguageServiceBrokerShim languageServiceBroker,
+            IThreadingContext threadingContext,
+            IAsynchronousOperationListener asyncListener,
+            IVsEditorAdaptersFactoryService editorAdaptersFactoryService,
+            IVsCodeWindow codeWindow)
+        {
+            InitializeComponent();
+
+            _languageServiceBroker = languageServiceBroker;
+            _threadingContext = threadingContext;
+            _asyncListener = asyncListener;
+            _editorAdaptersFactoryService = editorAdaptersFactoryService;
+            _codeWindow = codeWindow;
+            _cancellationTokenSource = new CancellationTokenSource();
+            _cancellationToken = _cancellationTokenSource.Token;
+            SortOption = SortOption.Location;
+
+            _computeDataModelQueue = new AsyncBatchingWorkQueue<bool, DocumentSymbolDataModel?>(
+                DelayTimeSpan.Short,
+                ComputeDataModelAsync,
+                EqualityComparer<bool>.Default,
+                asyncListener,
+                _cancellationToken);
+
+            _filterAndSortDataModelQueue = new AsyncBatchingWorkQueue<bool, DocumentSymbolDataModel?>(
+                DelayTimeSpan.NearImmediate,
+                FilterAndSortDataModelAsync,
+                EqualityComparer<bool>.Default,
+                asyncListener,
+                _cancellationToken);
+
+            _highlightExpandAndPresentItemsQueue = new AsyncBatchingWorkQueue<ExpansionOption>(
+                DelayTimeSpan.NearImmediate,
+                HighlightExpandAndPresentItemsAsync,
+                asyncListener,
+                _cancellationToken);
+
+            // We don't think the shell is initialized lazily, so we'll Debug.Fail(), but if it was we'd still
+            // see the view created later so this will still function.
+            if (ErrorHandler.Failed(codeWindow.GetPrimaryView(out var primaryTextView)))
+                Debug.Fail("GetPrimaryView failed during DocumentOutlineControl initialization.");
+
+            if (ErrorHandler.Failed(StartTrackingView(primaryTextView)))
+                Debug.Fail("StartTrackingView failed during DocumentOutlineControl initialization.");
+
+            if (ErrorHandler.Succeeded(codeWindow.GetSecondaryView(out var secondaryTextView)))
+            {
+                if (ErrorHandler.Failed(StartTrackingView(secondaryTextView)))
+                    Debug.Fail("StartTrackingView failed during DocumentOutlineControl initialization.");
+            }
+
+            var subjectBuffer = _trackedTextViews[primaryTextView].TextBuffer;
+            _textViewEventSource = new CompilationAvailableTaggerEventSource(
```

Comment:
CompilationAvailableTaggerEventSource is a *very* special event source. it's effectively used to say: hey... source generators may have run and added files to the compilation *that woudl change the data i present*.  So run me again after that so my information is accurate.

I'm guessing (but could be wrong) that for document symbols your symbols for a particular document would not be affected by source-generated docs elsewhere.  So you could just make a CompositionEventSource instead of a CompilationAvailableEventSource.  That will be much cheaper.